### ISSUE TITLE
Add ability for shield to add args

### ISSUE THREAD

**Issue by Akxe** [Sytten/nexus-shield#181]
Hi, I would like to use this library, but I still want to use the `crud` operations from `nexus-prisma-plugin`, is there a way to add input arguments to `crud` via shield? Or is there some other way?

```ts
export const ClientOffer = objectType({
	name: 'ClientOffer',
	definition(t) {
		const model = t.model('Offer');
		model.offerID();
		model.title({
          shield: ruleType({
            async resolve: (root, args, ctx) => {
              return (! await isLoginValid(args.login));
            },
          })
        });
		model.carriersOfOffer({
			pagination: false,
		});
		model.periodsOfOffer({
			pagination: false,
		});
	},
});

export const OfferTypes = [
	PeriodsOfOffer,
	ClientOffer,
	extendType({
		type: 'Query',
		definition(t) {
			t.crud.offer({
				alias: 'clientOffer',
				type: 'ClientOffer',
			});
		},
	}),
];
```

**Comment by Sytten** [comment]
Currently no because I inject my typing in the generated nexus file, I don't know if there are hooks for the generated prisma nexus file. I can check.

**Comment by Sytten** [comment]
According to the devs I talked to on the prisma slacks it should already be working.
```
I thought we made the plug-in field extensions pass-through in the t.model and t.crud method. Have you tried Sytten? This was previously asked so users could use the built-in authorization plug-in. I'm pretty sure we did it and that it should be working
```
I will try to test it myself once I have more time, but probably not this week.

**Comment by Akxe** [comment]
I will inform you if I would give it a try, but for now, I just have too much to do :/

**Comment by davekiss** [comment]
I am also looking at trying to get this working with `t.crud` but seems like it's not applying the rule:

```
export const collectionMutations = extendType({
  type: "Mutation",
  definition(t) {
    t.crud.createOneCollection({
      shield: isAuthenticated()
    });
  }
});
```

`isAuthenticated` is a simple check in context for a `user` param

```
export const isAuthenticated = generic(
  ruleType({
    resolve: (root, args, ctx) => {
      try {
        if (ctx.user === null) {
          return handleError(errors.notAuthenticated);
        }
        return true;
      } catch (e) {
        return e;
      }
    }
  })
);
```

🌲🌲🌲🌲🌲

**Comment by Sytten** [comment]
Ok I will give it a try so I can fix it. If the typing works then its most likely just the way that the config is fetched that is different. I did a major change yesterday to support the nexus 1.0.0 release so that might have broken the fetch too.

**Comment by Sytten** [comment]
It works as expected, I added a new example to help: https://github.com/Sytten/nexus-shield/tree/master/examples/prisma
Your problem @davekiss is that you seem to return an error instead of throwing it. This library is different from graphql-shield in that regard, either you return a boolean or you throw an error.
I will close the issue for now but feel free to comment if you need more help.

**Comment by davekiss** [comment]
You're right! Good catch, I updated this on my end and seems to be working as expected.

Thanks so much for your effort on this project!

**Comment by Akxe** [comment]
Thanks guys, but this change is a huge plus too, but the point of this issue is, that i would like to add is that i would like to be able to add arguments to the argument of query/mutation.

I want to be able to add `login` key to argument to the query/mutation where shield is defined.

---

I want my `isAuthenticated` function to be able to add arguments to given query/mutation.

**Comment by Sytten** [comment]
@davekiss glad it works
@Akxe this is out of scope for this plugin, I suggest creating another plugin for that. I think it would be fairly simple and could be executed after the shield to mutate the args.

**Comment by Akxe** [comment]
Would you add it if I made a pull request for it? I do really think that having a generic rule that would define an argument to be had would bring great benefits.

```ts
export const isAuthenticated = generic(
  ruleType({
    args: {
      login: inputObjectType({
        name: 'Login',
        // required: true || false,
        definition(t) {
          t.string('name', { required: true });
          t.string('password', { required: true });
        },
      }),
    },
    resolve: async (root, args, ctx) => {
      try {
        // Here we know, that args must have properties defined above.
        ctx.isLoggedIn = await validateAuth(args.login);
      } catch (e) {
        return e;
      }
    }
  })
);
```

**Comment by Sytten** [comment]
I need more context on why you would need such rule. I have never seen the username and password passed in the args except for a login resolver. I don't see why you would do that, the auth should be done before resolvers are called and injected in the context.
